### DISCUSSION TITLE
clear button on input

### DISCUSSION THREAD

**Discussion by alexander-densley** [shadcn-ui/ui#3274] [Ideas]
pretty self explanatory, but it would be nice for a prop to exist to add a clear button for input fields

**Comment by grafficmedia** [comment]
yeah, this is built-in behavior for inputs. type="search" should have an X in the right side by default. Shadcn should not be removing default web behavior like this without a way to add it back.

**Reply by HichOoMa** [reply]
I think it should be added in the type="file" also

**Reply by softwarebyze** [reply]
there is a clear button if u add `type="search"`
<img width="904" alt="image" src="https://github.com/user-attachments/assets/79d6a6b1-f9b7-47c5-a43d-274d9ea5a9c7">

the shadcn clear button in @geudrik's  comment below (https://github.com/shadcn-ui/ui/discussions/3274#discussioncomment-10054930) looks much better though
<img width="755" alt="image" src="https://github.com/user-attachments/assets/5d4fe736-5866-435d-b5ce-2b8bc85977ba">

**Comment by hsavit1** [comment]
how are people doing this now??

**Comment by geudrik** [comment]
> it would be nice for a prop to exist to add a clear button for input fields
Since most inputs are controlled, this would be difficult to do I would think. You can use the code below though to get what you're after. 

The button has an `onClick` handler that clears the state var (in this case, the global filter of the datatable), which in turn clears the input box.

The input box has an `onChange` handler to set it.

```tsx
                <div className="relative w-full max-w-sm">
                    <Input
                        placeholder="Filter..."
                        value={globalFilter ?? ''}
                        onChange={(event) =>
                            // table.getColumn("email")?.setFilterValue(event.target.value)
                            setGlobalFilter(String(event.target.value))
                        }
                        className="max-w-sm"
                    />
                    <Button
                        type="button"
                        variant="ghost"
                        size="icon"
                        className="absolute right-1 top-1/2 -translate-y-1/2 h-7 w-7 text-gray-500 hover:text-gray-900 dark:text-gray-400 dark:hover:text-gray-100"
                        onClick={() => {
                            setGlobalFilter('')
                        }}
                    >
                        <XIcon className="h-4 w-4"/>
                        <span className="sr-only">Clear</span>
                    </Button>
                </div>
```

**Comment by CoolSpot** [comment]
You can also use "asChild" to get browser's native clear button, like this:
```jsx
<CommandInput placeholder="Search library..." value={inputValue} onValueChange={setInputValue} asChild>
    <input type="search" />
</CommandInput>
```

<img width="731" height="325" alt="image" src="https://github.com/user-attachments/assets/6a53f509-d7d2-4da4-9277-825276c4c0bd" />
